### DISCUSSION TITLE
How can playing sequence images not affect the efficiency of other user interactions?

### DISCUSSION THREAD

**Discussion by Jason0247** [airsdk/Adobe-Runtime-Support#2894] [Q&A]
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/91530077/bb7ea649-9987-48aa-a334-2409ab7a06f0)

May I ask you a question
I have an application that keeps playing 0-n sequence images
At this point, if I am playing some animations, such as using TweenLite to create a Sprite animation, it will become very slow, and when I stop playing sequence images, it will become normal.
Is there any way to play sequence images without affecting the efficiency of other animation elements?

**I have noticed this issue:
Scenario 1: When playing a sequence image, displaying a Bitmap will not get stuck, but if I add one or two more Bitmaps to play the sequence image, it will be slow.
Scenario 2: This is also the issue with the first scenario. If I generate two applications, that is to say, I start two execs so that they do not get stuck, and each program is fast when it is independent.
I have a question, what is the reason for this? Is there any way to improve efficiency in an application?**

**Comment by MalacTheLittle** [comment]
You shoud use LoaderContext.imageDecodingPolicy set to "onLoad" to avoid stage freezing while loading image data, e.g.:

```
var lc:LoaderContext = new LoaderContext(false, ApplicationDomain.currentDomain, null);
lc.imageDecodingPolicy = "onLoad";
lc.allowCodeImport = true;
var imageLoader:Loader = new Loader();
imageLoader.contentLoaderInfo.addEventListener(Event.COMPLETE, imgLoaderComplete);
imageLoader.load(new` URLRequest(File.applicationDirectory.resolvePath("image1.jpg").nativePath), lc);
```

**Reply by Jason0247** [reply]
`	imgFile.nativePath = fileList[index];

	imgFileStream.open(imgFile, FileMode.READ);
	imgFileStream.readBytes(fileByte, 0, imgFileStream.bytesAvailable);
	
	//使用Loader 对象将图像文件二进制数据加载进来（可加载SWF、GIF、JPEG 或 PNG 格式的文件）;
	//使用Loader 是方便通过loader.contentLoaderInfo获得Bitmap对象
	//loader.unloadAndStop();
	//loader.unload();
	imgLoader.loadBytes(fileByte,lc);`
Is LoaderContext loaded asynchronously?
I used this method, but it still had the same effect.
I am using the File method to load images. Do you know if there is a more efficient way?

**Reply by MalacTheLittle** [reply]
It is stated in the `LoaderContext` documentation that "Decoding images **on load** rather than **on demand** can improve animation and user interface performance." ([imageDecodingPolicy](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/system/LoaderContext.html#imageDecodingPolicy))

If you're not modifying image bytearray, there's no need to load it with `FileStream` class. It is more convenient to load it directly with the `Loader` class as in above example.

However, if you need to modify image bytearray, to avoid freezing, you should open/load it asynchronously with `FileStream.openAsync`, e.g.:

```
var imageLoader:Loader = new Loader();

var file:File = File.applicationDirectory.resolvePath("image1.jpg");
var fs:FileStream = new FileStream();
fs.addEventListener(Event.COMPLETE, fsOpened);
fs.openAsync(file, "read");

function fsOpened(event:Event):void{
	var ba:ByteArray = new ByteArray();
	fs.readBytes(ba);
	ba.position = 0;
	fs.close();

	var lc:LoaderContext = new LoaderContext(false, ApplicationDomain.currentDomain, null);
	lc.imageDecodingPolicy = "onLoad";
	lc.allowCodeImport = true;
	imageLoader.contentLoaderInfo.addEventListener(Event.COMPLETE, imgLoaderComplete);
	imageLoader.loadBytes(ba, lc);
}

function imgLoaderComplete(event:Event):void{
	var bmp:Bitmap = Bitmap(event.target.content);
	bmp.smoothing = true;
	...
}
```
